src/core/qgsvectorlayerrenderer.cpp

```diff
@@ -152,10 +153,22 @@ bool QgsVectorLayerRenderer::render()
                                      .setFilterRect( requestExtent )
                                      .setSubsetOfAttributes( mAttrNames, mFields );
 
-  if ( !rendererFilter.isEmpty() && rendererFilter != "TRUE" )
+  const QgsFeatureFilterProvider* featureFilterProvider = mContext.featureFilterProvider();
+  if ( featureFilterProvider != NULL)
+  {
+    featureFilterProvider->filterFeatures( mLayer, &featureRequest );
+  }
+  if ( !rendererFilter.isNull() )
   {
-    featureRequest.setFilterExpression( rendererFilter );
     featureRequest.setExpressionContext( mContext.expressionContext() );
+    if ( featureRequest.filterExpression() == NULL )
+    {
+      featureRequest.setFilterExpression( rendererFilter );
+    }
+    else
+    {
+      featureRequest.setFilterExpression( QString( "%s AND %s" ).arg( rendererFilter ).arg( featureRequest.filterExpression()->expression() ) );
```

Comment:
this should be "(%s) AND (%s)"

 also

    featureRequest.setFilterExpression( QString( "(%s) AND (%s)" ).arg( rendererFilter, featureRequest.filterExpression()->expression() ) );

is faster and safer (rather than chaining the args)